File: scripts/apply-anti-detection.py
#!/usr/bin/env python3

"""
灵活的Playwright反检测修改脚本
- 基于配置文件进行变量替换
- 支持回滚和代码同步
- 生成patch文件便于版本管理
"""

import json
import os
import re
import shutil
import subprocess
import sys
from pathlib import Path
from typing import Dict, List

class AntiDetectionModifier:
    def __init__(self, config_path: str = "anti-detection-config.json"):
        self.root_path = Path(__file__).parent.parent
        self.config_path = self.root_path / config_path
        self.config = self._load_config()
        self.backup_dir = self.root_path / ".anti-detection-backup"
        
    def _load_config(self) -> Dict:
        """加载配置文件"""
        with open(self.config_path, 'r', encoding='utf-8') as f:
            return json.load(f)
    
    def scan_all_files(self) -> List[Path]:
        """扫描所有需要处理的文件"""
        target_files = []
        
        # 优先处理配置中指定的目标文件
        for target_file in self.config["file_patterns"]["target_files"]:
            file_path = self.root_path / target_file
            if file_path.exists():
                target_files.append(file_path)
        
        # 扫描其他相关文件
        search_dirs = [
            "packages/playwright-core/src",
            "packages/injected/src", 
            "packages/playwright-ct-core/src"
        ]
        
        for search_dir in search_dirs:
            dir_path = self.root_path / search_dir
            if dir_path.exists():
                for pattern in self.config["file_patterns"]["search_patterns"]:
                    if pattern.startswith("**/"):
                        pattern = pattern[3:]  # 移除 **/ 前缀
                    files = list(dir_path.glob(f"**/*{pattern}"))
                    target_files.extend(files)
        
        # 去重并过滤存在的文件
        return list(set([f for f in target_files if f.exists()]))
    
    def backup_original_files(self, target_files: List[Path]):
        """备份原始文件"""
        if self.config["sync_strategy"]["backup_original"]:
            print("📁 备份原始文件...")
            self.backup_dir.mkdir(exist_ok=True)
            
            for file_path in target_files:
                # 保持目录结构的备份
                rel_path = file_path.relative_to(self.root_path)
                backup_path = self.backup_dir / rel_path
                backup_path.parent.mkdir(parents=True, exist_ok=True)
                shutil.copy2(file_path, backup_path)
                print(f"  备份: {rel_path}")

    def apply_replacements(self, target_files: List[Path]):
        """应用变量替换"""
        print("🔧 应用反检测变量替换...")
        
        # 合并所有变量替换规则
        all_replacements = {}
        for category, mappings in self.config["variable_replacements"].items():
            all_replacements.update(mappings)
        
        modified_files = []
        
        for file_path in target_files:
            try:
                with open(file_path, 'r', encoding='utf-8') as f:
                    content = f.read()
                
                original_content = content
                replacements_made = []
                
                # 应用所有替换规则 - 简单的字符串替换
                for old_var, new_var in all_replacements.items():
                    if old_var in content:
                        content = content.replace(old_var, new_var)
                        replacements_made.append(f"{old_var} -> {new_var}")
                
                if content != original_content:
                    with open(file_path, 'w', encoding='utf-8') as f:
                        f.write(content)
                    
                    rel_path = file_path.relative_to(self.root_path)
                    modified_files.append(rel_path)
                    print(f"  ✅ 修改 {rel_path}: {', '.join(replacements_made)}")
            
            except Exception as e:
                rel_path = file_path.relative_to(self.root_path)
                print(f"  ❌ 处理失败 {rel_path}: {e}")
        
        return modified_files

    def create_patch(self, modified_files: List[Path]):
        """创建patch文件便于版本管理"""
        if self.config["sync_strategy"]["create_patch"]:
            print("📄 创建patch文件...")
            
            patch_content = []
            for file_path in modified_files:
                backup_path = self.backup_dir / file_path
                target_path = self.root_path / file_path
                
                if backup_path.exists():
                    # 生成单个文件的diff
                    result = subprocess.run([
                        "git", "diff", "--no-index", 
                        str(backup_path), str(target_path)
                    ], capture_output=True, text=True, cwd=self.root_path)
                    
                    if result.stdout:
                        patch_content.append(result.stdout)
            
            if patch_content:
                patch_file = self.root_path / "anti-detection.patch"
                with open(patch_file, 'w', encoding='utf-8') as f:
                    f.write('\n'.join(patch_content))
                print(f"  ✅ Patch文件: {patch_file}")

    def rollback(self):
        """回滚所有修改"""
        if not self.backup_dir.exists():
            print("❌ 没有找到备份目录")
            return
            
        print("🔄 回滚修改...")
        restored_count = 0
        
        # 遍历备份目录，恢复所有文件
        for backup_file in self.backup_dir.rglob("*"):
            if backup_file.is_file():
                # 计算原始文件路径
                rel_path = backup_file.relative_to(self.backup_dir)
                original_file = self.root_path / rel_path
                
                if original_file.exists():
                    shutil.copy2(backup_file, original_file)
                    print(f"  ✅ 恢复: {rel_path}")
                    restored_count += 1
        
        print(f"✅ 回滚完成，恢复了 {restored_count} 个文件")

    def validate_modifications(self, target_files: List[Path]):
        """验证修改是否成功"""
        print("🔍 验证修改结果...")
        
        all_replacements = {}
        for category, mappings in self.config["variable_replacements"].items():
            all_replacements.update(mappings)
        
        for file_path in target_files:
            try:
                with open(file_path, 'r', encoding='utf-8') as f:
                    content = f.read()
                
                rel_path = file_path.relative_to(self.root_path)
                
                # 检查原始变量是否还存在
                remaining_vars = []
                for old_var in all_replacements.keys():
                    if old_var in content:
                        remaining_vars.append(old_var)
                
                if remaining_vars:
                    print(f"  ⚠️  {rel_path} 中仍有原始变量: {', '.join(remaining_vars)}")
                
                # 检查新变量是否存在
                new_vars_found = []
                for new_var in all_replacements.values():
                    if new_var in content:
                        new_vars_found.append(new_var)
                
                if new_vars_found:
                    print(f"  ✅ {rel_path} 中找到新变量: {', '.join(new_vars_found)}")
            
            except Exception as e:
                rel_path = file_path.relative_to(self.root_path)
                print(f"  ❌ 验证失败 {rel_path}: {e}")

    def apply(self):
        """执行完整的修改流程"""
        print("🚀 开始应用Playwright反检测修改...")
        print(f"📝 配置: {self.config['description']}")
        
        # 1. 扫描所有文件
        target_files = self.scan_all_files()
        print(f"📁 发现 {len(target_files)} 个文件需要处理")
        
        # 2. 备份原始文件
        self.backup_original_files(target_files)
        
        # 3. 应用替换
        modified_files = self.apply_replacements(target_files)
        
        # 4. 创建patch
        self.create_patch(modified_files)
        
        # 5. 验证修改
        self.validate_modifications([self.root_path / f for f in modified_files])
        
        print(f"✅ 修改完成！共修改了 {len(modified_files)} 个文件")
        return modified_files

def main():
    if len(sys.argv) > 1:
        if sys.argv[1] == "rollback":
            modifier = AntiDetectionModifier()
            modifier.rollback()
            return
        elif sys.argv[1] == "validate":
            modifier = AntiDetectionModifier()
            target_files = modifier.scan_all_files()
            modifier.validate_modifications(target_files)
            return
    
    modifier = AntiDetectionModifier()
    modified_files = modifier.apply()
    
    print("\n📋 后续步骤:")
    print("1. npm run build  # 构建修改后的代码")
    print("2. bash scripts/build-custom-driver.sh  # 构建自定义driver包")
    print("3. python scripts/apply-anti-detection.py rollback  # 如需回滚")

if __name__ == "__main__":
    main()

File: scripts/apply-anti-detection.js
#!/usr/bin/env node

/**
 * Playwright反检测变量替换脚本
 * 基于 anti-detection-config.json 配置文件
 * 使用 Node.js 实现，支持 npm 运行
 */

const fs = require('fs');
const path = require('path');
const { glob } = require('glob');

class AntiDetectionModifier {
  constructor(configPath = 'anti-detection-config.json') {
    this.rootPath = path.resolve(__dirname, '..');
    this.configPath = path.join(this.rootPath, configPath);
    this.config = this.loadConfig();
    this.backupDir = path.join(this.rootPath, '.anti-detection-backup');
  }

  loadConfig() {
    try {
      const content = fs.readFileSync(this.configPath, 'utf-8');
      return JSON.parse(content);
    } catch (error) {
      console.error(`❌ 加载配置文件失败: ${error.message}`);
      process.exit(1);
    }
  }

  async scanAllFiles() {
    const targetFiles = new Set();
    
    // 优先处理配置中指定的目标文件
    for (const targetFile of this.config.file_patterns.target_files) {
      const filePath = path.join(this.rootPath, targetFile);
      if (fs.existsSync(filePath)) {
        targetFiles.add(filePath);
      }
    }
    
    // 扫描其他相关文件
    const searchDirs = [
      'packages/playwright-core/src',
      'packages/injected/src',
      'packages/playwright-ct-core/src'
    ];
    
    for (const searchDir of searchDirs) {
      const dirPath = path.join(this.rootPath, searchDir);
      if (fs.existsSync(dirPath)) {
        for (const pattern of this.config.file_patterns.search_patterns) {
          const cleanPattern = pattern.replace('**/', '');
          const globPattern = path.join(dirPath, '**', `*${cleanPattern.slice(1)}`);
          try {
            const files = await glob(globPattern);
            files.forEach(file => targetFiles.add(file));
          } catch (error) {
            console.warn(`⚠️  扫描模式失败 ${globPattern}: ${error.message}`);
          }
        }
      }
    }
    
    return Array.from(targetFiles).filter(file => fs.existsSync(file));
  }

  backupOriginalFiles(targetFiles) {
    if (this.config.sync_strategy.backup_original) {
      console.log('📁 备份原始文件...');
      
      if (!fs.existsSync(this.backupDir)) {
        fs.mkdirSync(this.backupDir, { recursive: true });
      }
      
      for (const filePath of targetFiles) {
        const relPath = path.relative(this.rootPath, filePath);
        const backupPath = path.join(this.backupDir, relPath);
        const backupDirPath = path.dirname(backupPath);
        
        if (!fs.existsSync(backupDirPath)) {
          fs.mkdirSync(backupDirPath, { recursive: true });
        }
        
        fs.copyFileSync(filePath, backupPath);
        console.log(`  备份: ${relPath}`);
      }
    }
  }

  applyReplacements(targetFiles) {
    console.log('🔧 应用反检测变量替换...');
    
    // 合并所有变量替换规则
    const allReplacements = {};
    for (const [category, mappings] of Object.entries(this.config.variable_replacements)) {
      Object.assign(allReplacements, mappings);
    }
    
    const modifiedFiles = [];
    
    for (const filePath of targetFiles) {
      try {
        const content = fs.readFileSync(filePath, 'utf-8');
        let modifiedContent = content;
        const replacementsMade = [];
        
        // 应用所有替换规则
        for (const [oldVar, newVar] of Object.entries(allReplacements)) {
          if (modifiedContent.includes(oldVar)) {
            modifiedContent = modifiedContent.replace(new RegExp(oldVar, 'g'), newVar);
            replacementsMade.push(`${oldVar} -> ${newVar}`);
          }
        }
        
        if (modifiedContent !== content) {
          fs.writeFileSync(filePath, modifiedContent, 'utf-8');
          
          const relPath = path.relative(this.rootPath, filePath);
          modifiedFiles.push(relPath);
          console.log(`  ✅ 修改 ${relPath}: ${replacementsMade.join(', ')}`);
        }
      } catch (error) {
        const relPath = path.relative(this.rootPath, filePath);
        console.error(`  ❌ 处理失败 ${relPath}: ${error.message}`);
      }
    }
    
    return modifiedFiles;
  }

  rollback() {
    if (!fs.existsSync(this.backupDir)) {
      console.error('❌ 没有找到备份目录');
      return;
    }
    
    console.log('🔄 回滚修改...');
    let restoredCount = 0;
    
    // 递归恢复所有备份文件
    const restoreFiles = (dir) => {
      const items = fs.readdirSync(dir);
      
      for (const item of items) {
        const backupPath = path.join(dir, item);
        const stat = fs.statSync(backupPath);
        
        if (stat.isDirectory()) {
          restoreFiles(backupPath);
        } else {
          const relPath = path.relative(this.backupDir, backupPath);
          const originalPath = path.join(this.rootPath, relPath);
          
          if (fs.existsSync(originalPath)) {
            fs.copyFileSync(backupPath, originalPath);
            console.log(`  ✅ 恢复: ${relPath}`);
            restoredCount++;
          }
        }
      }
    };
    
    restoreFiles(this.backupDir);
    console.log(`✅ 回滚完成，恢复了 ${restoredCount} 个文件`);
  }

  validateModifications(targetFiles) {
    console.log('🔍 验证修改结果...');
    
    const allReplacements = {};
    for (const [category, mappings] of Object.entries(this.config.variable_replacements)) {
      Object.assign(allReplacements, mappings);
    }
    
    for (const filePath of targetFiles) {
      try {
        const content = fs.readFileSync(filePath, 'utf-8');
        const relPath = path.relative(this.rootPath, filePath);
        
        // 检查原始变量是否还存在
        const remainingVars = [];
        for (const oldVar of Object.keys(allReplacements)) {
          if (content.includes(oldVar)) {
            remainingVars.push(oldVar);
          }
        }
        
        if (remainingVars.length > 0) {
          console.log(`  ⚠️  ${relPath} 中仍有原始变量: ${remainingVars.join(', ')}`);
        }
        
        // 检查新变量是否存在
        const newVarsFound = [];
        for (const newVar of Object.values(allReplacements)) {
          if (content.includes(newVar)) {
            newVarsFound.push(newVar);
          }
        }
        
        if (newVarsFound.length > 0) {
          console.log(`  ✅ ${relPath} 中找到新变量: ${newVarsFound.join(', ')}`);
        }
      } catch (error) {
        const relPath = path.relative(this.rootPath, filePath);
        console.error(`  ❌ 验证失败 ${relPath}: ${error.message}`);
      }
    }
  }

  async apply() {
    console.log('🚀 开始应用Playwright反检测修改...');
    console.log(`📝 配置: ${this.config.description}`);
    
    // 1. 扫描所有文件
    const targetFiles = await this.scanAllFiles();
    console.log(`📁 发现 ${targetFiles.length} 个文件需要处理`);
    
    // 2. 备份原始文件
    this.backupOriginalFiles(targetFiles);
    
    // 3. 应用替换
    const modifiedFiles = this.applyReplacements(targetFiles);
    
    // 4. 验证修改
    this.validateModifications(targetFiles.filter(f => {
      const relPath = path.relative(this.rootPath, f);
      return modifiedFiles.includes(relPath);
    }));
    
    console.log(`✅ 修改完成！共修改了 ${modifiedFiles.length} 个文件`);
    return modifiedFiles;
  }
}

// 主函数
async function main() {
  const args = process.argv.slice(2);
  
  if (args.includes('rollback')) {
    const modifier = new AntiDetectionModifier();
    modifier.rollback();
    return;
  }
  
  if (args.includes('validate')) {
    const modifier = new AntiDetectionModifier();
    const targetFiles = await modifier.scanAllFiles();
    modifier.validateModifications(targetFiles);
    return;
  }
  
  const modifier = new AntiDetectionModifier();
  const modifiedFiles = await modifier.apply();
  
  console.log('\n📋 后续步骤:');
  console.log('1. npm run build  # 构建修改后的代码');
  console.log('2. npm run build:driver  # 构建自定义driver包');
  console.log('3. npm run anti-detection:rollback  # 如需回滚');
}

// 运行
if (require.main === module) {
  main().catch(error => {
    console.error('❌ 执行失败:', error.message);
    process.exit(1);
  });
}

module.exports = { AntiDetectionModifier };